### PR TITLE
Export JsValue from js-sys

### PR DESCRIPTION
**Pull request by kajacx**
This way, you can name the `JsValue` type without adding a dependency on `wasm-bindgen`.

On the other hand `js-sys` already depends on `wasm-bindgen`, so it's not that bad. Still, returning a type (or requiring it as an argument) that your crate doesn't provide simply feels wrong.

### CODE REVIEW

**Comment by daxpedda** [comment]
I don't think I would mind that.
It would probably make sense to re-export `wasm-bindgen` itself as well.

Could you add this to the changelog as well?

File: crates/js-sys/CHANGELOG.md
Comment (daxpedda) [inline crates/js-sys/CHANGELOG.md:11]: If you rebase onto `main` you will find a good template to do this.
The text is fine, just missing the link to the PR.

File: crates/js-sys/src/lib.rs
Comment (daxpedda) [inline crates/js-sys/src/lib.rs:32]:
```suggestion
pub use wasm_bindgen::prelude::*;
```
I actually meant that we should re-export `wasm-bindgen` **in addition** to `JsValue`, but the prelude is even better!

File: crates/js-sys/src/lib.rs
Comment (daxpedda) [inline crates/js-sys/src/lib.rs:31]:
```suggestion
pub extern crate wasm_bindgen;
```
This is how reexports are usually done.

**Review by daxpedda — CHANGES_REQUESTED** [review]
> The downside is that unrelated parts of `wasm-bindgen` are exported as well (for example, you can annotate your methods with a `#[js_sys::wasm_bindgen::wasm_bindgen]` macro, which makes little sence, but going with a fine comb to reexport only the types used seems like a lot of work.

Reexports like this are just to make it easier to use dependencies without having to pull in sub-dependencies. So it's fine like this, it's a known pattern in Rust.

This is an opinionated change anyway, so we will wait for another maintainer to chime in.

File: crates/js-sys/src/lib.rs
Comment (kajacx) [reply crates/js-sys/src/lib.rs:31]: The only re-export I have worked was in [`wai-bindgen-wasmtime`](https://github.com/wasmerio/wai/blob/main/crates/wasmtime/src/lib.rs#L8) and [`wai-bindgen-wasmer`](https://github.com/wasmerio/wai/blob/main/crates/wasmer/src/lib.rs#L8). Also [this link](https://www.lurklurk.org/effective-rust/re-export.html) suggested using the `pub use dependency_crate` syntax.

But if you want `pub extern crate wasm_bindgen;`, go for it, it's your repo. I checked the "allow maintainers to make changes", so that means you can put the change into my fork of the repo? I'm not super familiar with this on GitHub.

**Comment by hamza1311** [comment]
If this is done, I don't think it should be done in just js-sys. Why not web-sys as well?

File: crates/js-sys/src/lib.rs
Comment (daxpedda) [reply crates/js-sys/src/lib.rs:31]: I recently found out that `pub extern` and `pub use` basically emit the same things on docs.rs.

**Comment by daxpedda** [comment]
I agree. WDYT overall though?

**Comment by hamza1311** [comment]
> WDYT overall though?

I'm not a fan of it. All it saves is one line in Cargo.toml. But I also can't think of a reason not to add it

**Comment by daxpedda** [comment]
It's generally considered good behavior to re-export crates that you use types of in your public API, which is the case for us in `js-sys` and `web-sys`.

But I'm happy to proceed now @kajacx if you finish this up.

File: crates/js-sys/src/lib.rs
Comment (daxpedda) [reply crates/js-sys/src/lib.rs:32]: I think we should leave it like this.

File: crates/js-sys/CHANGELOG.md
Comment (daxpedda) [inline crates/js-sys/CHANGELOG.md:11]:
```suggestion
* Re-export `wasm-bindgen` from `js-sys` and `web-sys`.
  [#3466](https://github.com/rustwasm/wasm-bindgen/pull/3466)

* Re-export `js-sys` from `web-sys`.
```

**Review by daxpedda — CHANGES_REQUESTED** [review]
Let's also re-export `wasm-bindgen` and `js-sys` from `web-sys`.

**Comment by kajacx** [comment]
> All it saves is one line in Cargo.toml

Yea, but tracking down that line is non-trivial. You have to go to the source of `js-sys` or find it on crates.io to get the version of `wasm-bindgen` it is using. And then you have to keep it updated as `js-sys` and it's `wasm-bindgen` dependency gets updated.

If you could use `js_sys::wasm_bindgen` instead, then you would have a 100% guarantee that you are using the same version of `wasm_bindgen` as `js_sys` is using, without the need to manually update anything.

On the other hand, if someone uses my crate, and *they* want to use `JsValue` as well, then they will you have to write `my_crate::js_sys::wasm_bindgen::JsValue`, which is getting ridiculous.

In the end, this is a preference decision, so I think it's best to keep it to the authors of the crate to make the final call. For example, I will not be exporting `js-sys` from my crate in this way, since that would defeat it's purpose.

**Comment by daxpedda** [comment]
> Let's also re-export `wasm-bindgen` and `js-sys` from `web-sys`.

@kajacx this is still missing, otherwise this is good to go.

**Comment by kajacx** [comment]
> > Let's also re-export `wasm-bindgen` and `js-sys` from `web-sys`.
> 
> @kajacx this is still missing, otherwise this is good to go.

Done.

**Comment by daxpedda** [comment]
Thanks!

**Pull request merged**
